pkg/utils/webhook/certificates.go

```diff
@@ -0,0 +1,185 @@
+func GenerateCertificates(ctx context.Context, mgr manager.Manager, certDir, namespace, name, url string) ([]byte, error) {
+	var (
+		caCert     *certificates.Certificate
+		serverCert *certificates.Certificate
+		err        error
+	)
+
+	// If the namespace is not set then the webhook controller is running locally. We simply generate a new certificate in this case.
+	if len(namespace) == 0 {
+		caCert, serverCert, err = generateNewCAAndServerCert(namespace, name, url)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error generating new certificates for webhook server")
+		}
+		return writeCertificates(certDir, caCert, serverCert)
+	}
+
+	// The controller stores the generated webhook certificate in a secret in the cluster. It tries to read it. If it does not exist a
+	// new certificate is generated.
+	c, err := getClient(mgr)
+	if err != nil {
+		return nil, err
+	}
+
+	secret := &corev1.Secret{}
+	if err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: certSecretName}, secret); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(err, "error getting cert secret")
+		}
+
+		// The secret was not found, let's generate new certificates and store them in the secret afterwards.
+		caCert, serverCert, err = generateNewCAAndServerCert(namespace, name, url)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error generating new certificates for webhook server")
+		}
+
+		secret.ObjectMeta = metav1.ObjectMeta{Namespace: namespace, Name: certSecretName}
+		secret.Type = corev1.SecretTypeOpaque
+		secret.Data = map[string][]byte{
+			certificates.DataKeyCertificateCA: caCert.CertificatePEM,
+			certificates.DataKeyPrivateKeyCA:  caCert.PrivateKeyPEM,
+			certificates.DataKeyCertificate:   serverCert.CertificatePEM,
+			certificates.DataKeyPrivateKey:    serverCert.PrivateKeyPEM,
+		}
+		if err := c.Create(ctx, secret); err != nil {
+			return nil, err
+		}
+
+		return writeCertificates(certDir, caCert, serverCert)
+	}
+
+	// The secret has been found and we are now trying to read the stored certificate inside it.
+	caCert, serverCert, err = loadExistingCAAndServerCert(secret.Data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading data of secret %s/%s", namespace, certSecretName)
+	}
+	return writeCertificates(certDir, caCert, serverCert)
+}
+
+func generateNewCAAndServerCert(namespace, name, url string) (*certificates.Certificate, *certificates.Certificate, error) {
+	caConfig := &certificates.CertificateSecretConfig{
+		CommonName: "webhook-ca",
+		CertType:   certificates.CACert,
+	}
+
+	caCert, err := caConfig.GenerateCertificate()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var (
+		dnsNames    []string
+		ipAddresses []net.IP
+	)
+
+	dnsNames = []string{
+		fmt.Sprintf("gardener-extension-%s", name),
+		fmt.Sprintf("gardener-extension-%s.%s", name, namespace),
+		fmt.Sprintf("gardener-extension-%s.%s.svc", name, namespace),
+	}
```

Comment:
these look like a copy paste issue?
in our case the `gardener-extension-` prefix can be omitted